modules/openxoutstreamBidAdapter.js

```diff
@@ -0,0 +1,224 @@
+
+  if (bids.some(bid => bid.params.doNotTrack)) {
+    queryParams.ns = 1;
+  }
+
+  if (bids.some(bid => bid.params.coppa)) {
+    queryParams.tfcd = 1;
+  }
+
+  let url = `https://${bids[0].params.delDomain}/v/1.0/avjp`
+  return {
+    method: 'GET',
+    url: url,
+    data: queryParams,
+    payload: {'bids': bids}
+  };
+}
+
+function handleVastResponse(response, serverResponse) {
+  const body = response.body
+  let bidResponses = [];
+  if (response !== undefined && body.vastUrl !== '' && body.pub_rev && body.pub_rev > 0) {
+    let vastQueryParams = parse(body.vastUrl).search || {};
+    const openHtmltag = '<html><head></head><body>';
+    const closeHtmlTag = '</body></html>';
+    const sdkScript = createSdkScript().outerHTML;
+    const placementDiv = createPlacementDiv();
+    placementDiv.dataset.pID = PUBLISHER_ID;
+    const placementDivString = placementDiv.outerHTML;
+    const adResponse = getTemplateAdResponse(body.vastUrl, PLACEMENT_ID);
+    const adResponseString = JSON.stringify(adResponse);
+    const ymAdsScript = '<script type="text/javascript"> window.__ymAds =' + adResponseString + '</script>';
+
+    let bidResponse = {};
+    bidResponse.requestId = serverResponse.bids[0].bidId;
+    bidResponse.bidderCode = BIDDER_CODE;
+    bidResponse.netRevenue = NET_REVENUE;
+    bidResponse.currency = CURRENCY;
+    bidResponse.cpm = Number(body.pub_rev) / 1000;
+    bidResponse.creativeId = body.adid;
+    bidResponse.height = body.height;
+    bidResponse.width = body.width;
+    bidResponse.vastUrl = body.vastUrl;
+    bidResponse.ttl = TIME_TO_LIVE;
+    bidResponse.mediaType = BANNER;
+    bidResponse.ad = openHtmltag + placementDivString + ymAdsScript + sdkScript + closeHtmlTag;
+
+    response.ph = vastQueryParams.ph;
+    response.colo = vastQueryParams.colo;
+    response.ts = vastQueryParams.ts;
+    bidResponses.push(bidResponse);
+  }
+  return bidResponses;
+}
+registerBidder(spec);
+
+// HELPER FUNCTIONS
+function createSdkScript() {
+  const script = document.createElement('script');
+  script.innerHTML = YM_SCRIPT;
+  return script;
+}
+function createPlacementDiv() {
+  const div = document.createElement('div');
+  div.id = `ym_${PLACEMENT_ID}`;
+  div.classList.add('ym');
+  div.dataset.lfid = `${LICENSED_FORMAT_ID}`;
+  return div
+}
+
+/**
+ * Create a nativeplay template with the placement id and vastURL.
+ * @param vastUrl
+ * @param placementId
+ */
+const getTemplateAdResponse = (vastUrl, placementId) => {
+  return {
+    availability_zone: 'us-east-1a',
+    data: [
+      {
+        ads: [
+          {
+            actions: {},
+            ad_id: 1991358644725162800,
+            adv_id: '1991358644725162800',
+            configurables: {
+              cta_button_copy: 'Learn More',
+              vast_click_tracking: 'true',
+              vast_url: vastUrl,
+            },
+            cr_id: '1991358644725162817',
```

Comment:
@HolzAndrew , let's use the constants declared above to fill in the `cr_id` and other values.

Let's move `ad_id` to a constant as well, and use that to replace both `ad_id` and `adv_id` in the template.

Reply:
makes sense. done.